Add drivers and shields for Arduino Modulino Buttons and Smart LEDs

fabiobaltieri opened the pull request:
Add drivers and shields for Arduino Modulino Buttons and Smart LEDs, these are QWIIC modules to provide buttons and LED strip functionalities.

fabiobaltieri committed:
input: add arduino,modulino-buttons

Add an input driver for the modulino buttons module. This is a pluggable
I2C board implementing three buttons and three LEDs, the I2C protocol is
implemented on a microcontroller on the modulino board itself, the
firmware for that is open source and can be updated using an Arduino
sketch:

Link: https://github.com/arduino/node_modulino_firmware
Link: https://github.com/arduino-libraries/Modulino
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
leds: add arduino,modulino-buttons-leds

Add an LED driver for the modulino buttons module, this has three LEDs
that are controllable independently of the buttons.

Link: https://github.com/arduino/node_modulino_firmware
Link: https://github.com/arduino-libraries/Modulino
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
leds: add arduino,modulino-smartleds

Add an led_strip driver for the modulino smartleds module. This is a
pluggable I2C board with 8 addressable RGB LEDs

The I2C protocol is implemented on an microcontroller on the modulino
board itself, the firmware for that is open source and can be updated
using an Arduino sketch:

Link: https://github.com/arduino/node_modulino_firmware
Link: https://github.com/arduino-libraries/Modulino
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
shields: add two shield definitions for modulino boards

Add two shield definitions for the the modulino buttons and smartleds
modules.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

kartben commented on boards/shields/arduino_modulino_buttons/doc/img/arduino_modulino_buttons.png:
Run the images through cwebp if you have a chance, will probably half the size

fabiobaltieri replied:
sound, done

kartben commented on dts/bindings/mfd/arduino,modulino-buttons-common.yaml:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: Arduino Modulino Buttons (common)
+
```
This binding named "-common" that ends up being used in the DT directly instead of being included in another binding kind of confuses/bothers me

Either name arduino,modulino-buttons and then the buttons `arduino,modulino-buttons-buttons`, or just use `i2c-device` directly in the shield overlay?

fabiobaltieri replied:
yeah good idea, no point keeping it went for i2c-device, dropped the commit

kartben commented on drivers/led_strip/Kconfig.arduino:
```diff
@@ -0,0 +1,10 @@
+	depends on DT_HAS_ARDUINO_MODULINO_SMARTLEDS_ENABLED
+	select I2C
+	help
+	  Enable driver Arduino Modulino Buttons LEDs.
```
👀

kartben commented on drivers/led/modulino_buttons_leds.c:
```diff
@@ -0,0 +1,107 @@
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(modulino_buttons_leds, CONFIG_INPUT_LOG_LEVEL);
```
```suggestion
LOG_MODULE_REGISTER(modulino_buttons_leds, CONFIG_LED_LOG_LEVEL);
```

... thank you copilot 🫣

fabiobaltieri replied:
AI AI

soburi commented on boards/shields/arduino_modulino_buttons/Kconfig.shield:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_ARDUINO_MODULINO_BUTTONS
+	def_bool $(shields_list_contains,rpi_pico_uno_flexypin)
```
Copy-paste mistake?

fabiobaltieri replied:
yeah I'll put it on the ever growing list of copy paste messup I did on this one

fabiobaltieri commented on boards/shields/arduino_modulino_smartleds/Kconfig.shield:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_ARDUINO_MODULINO_BUTTONS
+	def_bool $(shields_list_contains,rpi_pico_uno_flexypin)
```
name is wrong too -_-

fabiobaltieri replied:
done

fabiobaltieri commented on drivers/led/modulino_buttons_leds.c:
```diff
@@ -0,0 +1,107 @@
+static int modulino_buttons_leds_set_brightness(const struct device *dev,
+						uint32_t led, uint8_t value)
+{
+	return modulino_buttons_leds_set(dev, led, value > 0);
```
interesting, the problem is that one could use the set brightness API and have it working on both pwm and normal LEDs, if I don't implement it that would not work, but then this would be better handled in the set brightness LED implementation

I'll leave it here for now if you don't mind and follow up with an API clean uproposal

fabiobaltieri replied:
Not sure either but https://github.com/zephyrproject-rtos/zephyr/pull/89718 is landing in your inbox soon

fabiobaltieri commented on boards/shields/arduino_modulino_buttons/Kconfig.shield:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_ARDUINO_MODULINO_BUTTONS
+	def_bool $(shields_list_contains,rpi_pico_uno_flexypin)
```
done

fabiobaltieri commented on drivers/input/input_modulino_buttons.c:
```diff
@@ -0,0 +1,99 @@
+		goto out;
+	}
+
+	LOG_HEXDUMP_DBG(buf, sizeof(buf), buf);
```
yeah let's get rid of it

fabiobaltieri commented on dts/bindings/input/arduino,modulino-buttons.yaml:
```diff
@@ -0,0 +1,20 @@
+    required: true
+    description: |
+      Key codes to emit, the module has three buttons so this must specify
+      three code.
```
ok boss

fabiobaltieri commented on boards/shields/arduino_modulino_smartleds/Kconfig.shield:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_ARDUINO_MODULINO_BUTTONS
+	def_bool $(shields_list_contains,rpi_pico_uno_flexypin)
```
one more one more!

fabiobaltieri commented:
@kartben @soburi thanks for the review, addressed all comments but the set_brightness one, I think that could use some API rework, looks like I rebased on main at some stage so I messed the diff on github, apologies for that

simonguinot reviewed (commented):
Hi @fabiobaltieri,

Please see below for some questions and comments I have regarding the LED and strip drivers.

fabiobaltieri commented on drivers/led/modulino_buttons_leds.c:
```diff
@@ -0,0 +1,107 @@
+		return -ENODEV;
+	}
+
+	/* Reset to all LEDs off */
```
go away

fabiobaltieri commented on drivers/led/modulino_buttons_leds.c:
```diff
@@ -0,0 +1,107 @@
+
+LOG_MODULE_REGISTER(modulino_buttons_leds, CONFIG_LED_LOG_LEVEL);
+
+#define MODULINO_BUTTONS_NUM_LEDS 3
```
no, the buttons led thing is just three single color leds

fabiobaltieri commented on drivers/led/modulino_buttons_leds.c:
```diff
@@ -0,0 +1,107 @@
+	}
+
+	/* Reset to all LEDs off */
+	ret = i2c_write_dt(&cfg->bus, data->buf, sizeof(data->buf));
```
that's the problem, there's no reset, hard or soft, so at this point it'll be zero if it's the first power off, but it's going to be whatever was left if the device restarted, so this write is going to be effectively a soft reset by clearing the current status

simonguinot replied:
OK.

It reminds me that we need a mechanism to apply a LED configuration or keep the current state (allowing to keep the same LED state or hardware effect across a reboot / reset) during the initialization.

fabiobaltieri commented on drivers/led/modulino_buttons_leds.c:
```diff
@@ -0,0 +1,107 @@
+
+struct modulino_buttons_leds_data {
+	uint8_t buf[MODULINO_BUTTONS_NUM_LEDS];
+};
```
can't, the i2c protocol does not allow addressing the individual bytes (note that I'm using `i2c_write_dt` with no register) so I need to persist the state in the driver to preserve the state of the other LEDs, don't love it, but this how @facchinm felt that day

simonguinot replied:
Is that something specific with the STM32C0 handling the buttons / LED ?

fabiobaltieri replied:
> Is that something specific with the STM32C0 handling the buttons / LED ?

No just how it's been implemented (https://github.com/arduino/node_modulino_firmware/blob/main/Core/Src/main.c), the C0 i2c target could have handled it fine.

> It is a shame to have an MCU to control the LEDs and not implement a register map to allow to address them individually :(

Yeah, the C0 firmware can be updated over the bus and the Arduino folks have a sketch for it, so theoretically this could be changed, but then there's devices in the wild with the current protocol, we'd have to provide some code to do the upgrade from zephyr as well etc...

But yeah missed opportunity.

fabiobaltieri commented on drivers/led_strip/modulino_smartleds.c:
```diff
@@ -0,0 +1,110 @@
+				MODULINO_SMARTLEDS_FULL_BRIGHTNESS |
+				(pixels[i].b << 16) |
+				(pixels[i].g << 8) |
+				pixels[i].r);
```
looks so, thanks for the pointer had no clue what LEDs it was, should be 8x of those driven from an STM32C0, added few notes about the protocol, renamed the marker thing as that's how it seems to be referred in the datasheet

fabiobaltieri commented on drivers/led_strip/modulino_smartleds.c:
```diff
@@ -0,0 +1,110 @@
+	for (uint8_t i = 0; i < count; i++) {
+		data->buf[i] = sys_cpu_to_be32(
+				MODULINO_SMARTLEDS_DEFAULT |
+				MODULINO_SMARTLEDS_FULL_BRIGHTNESS |
```
cool, added a note in the definitions on the top, about the overlapping byte, unfortunately I wrote this stuff quite some ago and I can't recall where I copied this from, but I suspect the reason for it is something on the line of "I screwed up".

fixed

fabiobaltieri commented:
Fixed a wrong bit in the protocol and added few comments, apart from that I think this is good for another round of review, lmk if I missed any comment, working on this on and off and between humans and computers I may as well have missed your comments. Thanks for the reviews, sorry for the copy paste madness. :-)

simonguinot reviewed (commented):
I wonder if we should say explicitly "somewhere" that these LED drivers are not for a firmware running on the Modulino board, but on an MCU connected to the Modulino. Or is it obvious ?

fabiobaltieri commented:
> I wonder if we should say explicitly "somewhere" that these LED drivers are not for a firmware running on the Modulino board, but on an MCU connected to the Modulino. Or is it obvious ?

Uh, not following, I may have explained something incorrectly: the driver are for a firmware running on the C0 mounted on the modulino board. As in: the modulino board is the connectable i2c board with buttons, leds or whatever and from our point of view it's just an i2c device. Some just have the actual chip, some have a C0 implementing the i2c target protocol.

simonguinot commented:
> > I wonder if we should say explicitly "somewhere" that these LED drivers are not for a firmware running on the Modulino board, but on an MCU connected to the Modulino. Or is it obvious ?
> 
> Uh, not following, I may have explained something incorrectly: the driver are for a firmware running on the C0 mounted on the modulino board. As in: the modulino board is the connectable i2c board with buttons, leds or whatever and from our point of view it's just an i2c device, it has the C0 AND the buttons/LEDs on board. Some just have the actual chip, these two (and others) have a C0 implementing the i2c target protocol.

Oh I got it all wrong ! I though that the drivers were for a firmware running on a mother board / MCU connected the Modulino daughter board (with the C0, button and LEDs). And that the C0 and its stock firmware was exposing an I2C device, allowing to control the LEDs and buttons...

OK. So, forget this comment. I have a new one, which I hope won't be as dumb :) If the `modulino_smartleds` driver directly drives an LC8822 LED strip, shouldn't it be called `lc8822` ?

kartben commented:
> Oh I got it all wrong ! I though that the drivers were for a firmware running on a mother board / MCU connected the Modulino daughter board (with the C0, button and LEDs). And that the C0 and its stock firmware was exposing an I2C device, allowing to control the LEDs and buttons...

Yes, I think that's exactly what it is 🫣 I think @fabiobaltieri may have been using confusing terms when saying it's a "driver for a firmware running on the C0", maybe. (Or I, too, I am getting it all wrong! :D)

fabiobaltieri commented:
Dammit I keep making this worse and worse, hope at least I'm confusing copilot too to make it say something funny.

So the C0 is on the modulino board, the three drivers in this PR are on any MCU connected to the Modulino board via I2C, the modulino boards, at least these ones, have an STM32C0 implementing a custom I2C protocol (that does not support register addressing), that firmware can be updated to modify the protocol but it seems like a big ask.

For the smartleds specifically, those LEDs are driven via SPI, the C0 has a firmware that is just taking the I2C data from our host, adding the synchronization bytes and then passing the actual LED data through for 8 LEDs, so there's some LED implementation leaking through but I'd say not enough to justify trying to reworking this into a generic `lc8822` driver, you'd basically only share one line of code and two defines.

simonguinot commented:
> Dammit I keep making this worse and worse, hope at least I'm confusing copilot enough to make it say something funny.
> 
> So the C0 is on the modulino board, the three drivers in this PR are on any MCU connected to the Modulino board via I2C, the modulino boards, at least these ones, have an STM32C0 implementing a custom I2C protocol (that does not support register addressing), that firmware can be updated to modify the protocol but it seems like a big ask.

OK ! That's how I understood it at first. So I am reinstalling my first comment: should we say somewhere (commit description, Kconfig) that these drivers are for any MCU connected to the Modulino board, and not the C0 ? Or is it obvious ? I honestly think there is room for confusion here :)

fabiobaltieri commented:
> OK ! That's how I understood it at first. So I am reinstalling my initial comment: should we say somewhere (commit description, Kconfig) that these drivers are for any MCU connected to the Modulino board, and not the C0 ? Or is it obvious ? I honestly think there is room for confusion here :)

Wouldn't say it's obvious, but I would reckon that users of this would just see them like any other i2c device without thinking too much as to how the protocol is implemented. That said, the firmware is open source and the code was the reference I used for the protocol, so yeah I think you are right that adding a reference in the commit message is a good idea, let me rework those.

fabiobaltieri commented:
Okay same code, added some more details in the commit messages and references to the library, firmware and a mention to the fact that they can be updated. That was a good comment @simonguinot, unfortunately GitHub tend to hide commit messages in a tab and for how much I'm preaching about how important they are I realize I'm tending to write poor ones myself.

simonguinot reviewed (dismissed):
The LED drivers are looking good to me !

simonguinot commented:
Thanks for updating the commit message. I think it is clearer.

soburi commented on drivers/led/Kconfig:
```diff
@@ -27,6 +27,7 @@ config LED_SHELL
 	  Enable LED shell for testing.
 
 # zephyr-keep-sorted-start
+source "drivers/led/Kconfig.arduino"
```
Since this is mostly "product name", `Kconfig.modulino` might be better.
It's not blocking.

fabiobaltieri replied:
Good observation, changed it, matches the file naming I chose as well, added that path to the Arduino area in the maintainers file as well while at it, I missed that too.

fabiobaltieri committed:
MAINTAINERS: add modulino files to the Arduino area

Add modulino drivers path to the Arduino area.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
> Issues ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [0 New issues](https://sonarcloud.io/project/issues?id=zephyrproject-rtos_zephyr&pullRequest=89685&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) ![](https://camo.githubusercontent.com/86e5c40907bf9e8933fe6bcf185f7076622a76a3bef4b6d2b5fd583959409aef/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f61636365707465642d313670782e706e67) [0 Accepted issues](https://sonarcloud.io/project/issues?id=zephyrproject-rtos_zephyr&pullRequest=89685&issueStatuses=ACCEPTED)
> 
> Measures ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=zephyrproject-rtos_zephyr&pullRequest=89685&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [0.0% Coverage on New Code](https://sonarcloud.io/component_measures?id=zephyrproject-rtos_zephyr&pullRequest=89685&metric=new_coverage&view=list) ![](https://camo.githubusercontent.com/187ba88c264795e888bdd18141ef29390601e5110d4e9c8cc48f2f8afc183ba0/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67) [0.0% Duplication on New Code](https://sonarcloud.io/component_measures?id=zephyrproject-rtos_zephyr&pullRequest=89685&metric=new_duplicated_lines_density&view=list)

Hey @sonarqubecloud, if you don't have anything to say, could you just not say anything?

pillo79 reviewed (approved):
Thanks Fabio!
